File: BetESS_a69303_pg25291 11nov_observer/BetESS/src/betess/MenuApostador.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package betess;

import java.util.Locale;
import java.util.Scanner;

/**
 *
 * @author ana almeida
 */
public class MenuApostador {

    public Scanner in = new Scanner(System.in).useLocale(Locale.US);
    Apostador user;
    private BetESS bd;
    private boolean flag;

    public MenuApostador(Apostador u, BetESS bd) {
        this.user = u;
        this.bd = bd;
        flag = true;

    }


    public void start() {
        while (flag) {
            System.out.println("************************************************");
            System.out.println("           Bem Vindo " + user.getNome());
            System.out.println("************************************************");
            System.out.println("Escolha a opção:");
            System.out.println("1-Listar Apostas");
            System.out.println("2-Apostar");
            System.out.println("3-Ganho em apostas fechadas");
            System.out.println("4-Valor disponivel para apostar");
            System.out.println("5-Listar Notificações");
            System.out.println("6-Limpar Notificações");
            System.out.println("7-Sair");
            String opt = in.next();

            switch (opt) {
                case ("1"):
                    System.out.println("************************************************");
                    menuListarApostas();
                    break;
                case ("2"):
                    System.out.println("************************************************");
                    menuApostar();
                    break;
                case ("3"):
                    System.out.println("************************************************");
                    menuGanhoApostas();
                    break;
                case ("4"):
                    System.out.println("************************************************");
                    System.out.println("Valor: " + this.user.getValor());
                    break;
                case ("5"):
                    System.out.println("************************************************");
                    listarNotificacoesOdd();
                    break;
                case ("6"):
                    System.out.println("************************************************");
                    limparNofificacoes();
                    break;
                case ("7"):
                    System.out.println("************************************************");
                    this.bd.updateUser(user);
                    flag = false;
                    break;

            }
        }

    }

    private void menuListarApostas() {
        for (int d : this.bd.getEventos().keySet()) {
            Evento e = this.bd.getEventos().get(d);
            if (e.getEstado()) {
                System.out.println(e.toString());
            }
        }
    }

    private void menuApostar() {

        System.out.println("Id da aposta");
        int idEvento = in.nextInt();
        if(this.bd.getEventos().get(idEvento).getEstado()){
        System.out.println("Qual o resultado a apostar:");
        System.out.println("0-Empate");
        System.out.println("1-Equipa 1");
        System.out.println("2-Equipa 2");
        int aposta = in.nextInt();
        System.out.println("Valor a apostar");
        double valor = in.nextDouble();
        if (this.user.getValor() < valor) {
            System.out.println("Não tem valor disponivel para essa aposta");
        } else {

            double v = this.user.getValor();
            this.user.setValor(v - valor);
            Evento e = this.bd.getEventos().get(idEvento);
            e.addObserver(user);

            //Aposta(int id, String eqA, String eqB, double oddA, double oddB, double oddEmp)
            Odd o = e.getOdd();
            Aposta a = new Aposta(this.user.getContadorAposta(), idEvento, aposta, valor, o.getOddA(), o.getOddB(), o.getOddEmpate());

            //System.out.println("APOSTA   A ################ "+ a.getOdd().getOddA());
            //a.setOdd(e.getOdd());
            //a.setValor(valor);
            //((Apostador)this.bd.getUsers().get(user.getEmail())).addAposta(idEvento, a);
            this.user.addAposta(a.getId(), a);
            System.out.println("Registado");
        }
        }
        else
            System.out.println("Aposta fechada");

    }

    private void menuGanhoApostas() {

        if (this.user.getListaApostas().isEmpty()) {
            System.out.println("Não existem apostas");
            System.out.println("");
        } else {
            for (int id : this.user.getListaApostas().keySet()) {

                if (this.user.getListaApostas().get(id).fechada()) {
                    Aposta a = user.getListaApostas().get(id);
                    //System.out.println("Aposta :" + a.getId());
                    System.out.println(this.bd.getEventos().get(id).toString());
                    System.out.println("Ganho : " + a.getGanho());
                }

            }
        }

    }

    private void limparNofificacoes() {
        this.user.limpaNoficacoesOdd();
    }

    private void listarNotificacoesOdd() {
        if (user.getNotificacoesOdd().isEmpty()) {
            System.out.println("Não existem alterações");
        } else {
            System.out.println("Novas odds: ");
            for (int idEvento : user.getNotificacoesOdd()) {
                System.out.println("EVENTO: " + this.bd.getEventos().get(idEvento).toString());
            }
        }
    }

}

File: Ruby_Observer/menu_apostador.rb
require_relative 'apostador'
require_relative 'aposta'
require_relative 'BetESS'
require_relative 'evento'
require 'time'


class MenuApostador

  def initialize(apostador, bet_ess)
    @bet_ess=bet_ess
    @apostador=apostador
    @flag = true
  end

  def start
    while @flag do
      p '************************************************'
      p '*                  APOSTADOR ' + "#{@apostador.nome}" + '              *'
      p '************************************************'
      p 'Escolha a opção'
      p '1-Listar Apostas'
      p '2-Apostar'
      p '3-Ganho em apostas fechadas'
      p '4-Valor disponível para apostar'
      p '5-Listar Notificações'
      p '6-Limpar Notificações'
      p '7-Sair'
      opt = gets.chomp

      case opt
        when '1' then
          p '************************************************'
          listar_apostas
        when '2' then
          p '************************************************'
          fazer_aposta
        when '3' then
          p '************************************************'
          ganho_apostas
        when '4' then
          p '************************************************'
          p 'Valor: ' + "#{@apostador.valor}"
        when '5' then
          p '************************************************'
          listar_notificacoes_odd
        when '6' then
          p '************************************************'
          limpar_nofificacoes
        when '7' then
          p '************************************************'
          @flag = false

      end
    end
  end

  def listar_apostas
    eventos = @bet_ess.get_eventos
    eventos.each do |evento|
      if evento.estado
        p "#{evento.to_s}"
      end
    end
  end

  def fazer_aposta
    p 'Id da aposta'
    id = gets.chomp.to_i
    if @bet_ess.exist_evento(id)
      p 'Qual o resultado a apostar:'
      p '0 - Empate'
      p '1 - Equipa 1'
      p '2 - Equipa 2'
      aposta = gets.chomp
      p 'Valor a apostar:'
      valor = gets.chomp.to_f
      if @apostador.valor.to_f < valor
        p 'Não tem valor disponível para essa aposta'
      else
        @apostador.set_valor(@apostador.valor.to_f - valor)
        @bet_ess.get_eventos[id].add_observer(@apostador)
        ap = Aposta.new(@apostador.contador_id_aposta, id, aposta, valor, @bet_ess.get_eventos[id].odd.odd_v, @bet_ess.get_eventos[id].odd.odd_d, @bet_ess.get_eventos[id].odd.odd_e, nil)
        @apostador.add_aposta(ap)
        p 'Registado'
      end
    else
      p 'Aposta fechada'
    end
  end

  def ganho_apostas
    if @apostador.lista_apostas.empty?
      p 'Não existem Apostas'
    else
      @apostador.lista_apostas.values.each do |aposta|
        if aposta.fechada
          ap = aposta
          p 'Ganho: ' + "#{ap.ganho}"
        end
      end
    end
  end

  def listar_notificacoes_odd

    if @apostador.not_odd.empty?
      p 'Não existem alterações'
    else
      p ' Novas odds nos seguintes eventos:'
      @apostador.not_odd.each do |idevento|
        p " #{@bet_ess.get_eventos[idevento].to_s}"

      end

    end
  end

  def limpar_nofificacoes
    @apostador.limpa_noficacoes_odd
  end

end

